Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -211,9 +211,31 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
     Cookie cookie;
     cookie.name = WTFMove(options.name);
     cookie.value = WTFMove(options.value);
-    cookie.domain = options.domain.isNull() ? document.domain() : WTFMove(options.domain);
     cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
 
+    cookie.domain = options.domain.isNull() ? document.domain() : WTFMove(options.domain);
+    auto& url = document.url();
+    if (!cookie.domain.isNull()) {
+        if (cookie.domain.startsWith('.')) {
+            promise->reject(Exception { TypeError, "The domain must not begin with a '.'"_s });
+            return;
+        }
+
+        auto host = url.host();
+        if (host != cookie.domain) {
+            if (host.length() <= cookie.domain.length() || host.characterAt(host.length() - cookie.domain.length() - 1) != '.' || !host.endsWith(cookie.domain)) {
```

Comment:
I'd write this with something like this:

if (!host.endsWith(cookie.domain) || !StringView(host).substring(0, host.length() - cookie.domain.length()).endsWith('.'))

Reply:
I have two questions about this:

1. Doesn't `StringView::substring()` allocate a new `StringView`? Which is what we were trying to avoid?
2. If `host == cookie.domain`, then this if condition will be true because '!host.endsWith(cookie.domain)` will be false and `!StringView(host).substring(0, host.length() - cookie.domain.length()).endsWith('.')` will be true. So we will end up rejecting when we shouldn't correct?

If `StringView::substring()` does not allocate a new `StringView`, then maybe we can use something like this: `if (!host.endsWith(cookie.domain) || (host.length() > cookie.domain.length() && !StringView(host).substring(0, host.length() - cookie.domain.length()).endsWith('.')))`. Otherwise, maybe there is some other more concise way?

Reply:
A StringView doesn't allocate anything.  It's just a few integers in registers.  makeString does allocate and copy.

You're right, you'll need some logic for if the strings are equal.  At the end of the day we're just talking about clean ways of writing the same thing.  If there's no way it can go out of bounds, they're all equivalent.

Reply:
Ok that sounds good, I've updated the code to use `StringView::substring()` with the extra check as I wrote above.